File: FirebaseApp (Healthia)/app/src/main/java/com/example/firebaseapp/activitys/LoginActivity.java
package com.example.firebaseapp.activitys;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.util.Patterns;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.example.firebaseapp.R;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.iid.FirebaseInstanceId;

import java.util.HashMap;

public class LoginActivity extends AppCompatActivity {

    //views
    TextInputEditText mEmailEt, mPasswordEt;
    TextView notHaveAccountTv;
    Button mLoginBtn;

    //Declare an instance of FirebaseAuth
    private FirebaseAuth mAuth;
    private DatabaseReference mUserDatabase;

    //progress dialog
    ProgressDialog pd;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        //actionbar and it's title
        ActionBar actionBar = getSupportActionBar();
        actionBar.setTitle("Login");

        //enable back button
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setDisplayShowHomeEnabled(true);

        //In the onCreate() method, initialize the FirebaseAuth instance.
        mAuth = FirebaseAuth.getInstance();
        mUserDatabase = FirebaseDatabase.getInstance().getReference().child("Users");

        //init
        mEmailEt = findViewById(R.id.emailEt);
        mPasswordEt = findViewById(R.id.passwordEt);
        notHaveAccountTv = findViewById(R.id.not_have_accountTv);
        mLoginBtn = findViewById(R.id.login_btn);

        //login button click
        mLoginBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                //input data
                String email = mEmailEt.getText().toString().trim();
                String password = mPasswordEt.getText().toString().trim();
                if (!Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
                    //invalid email pattern set error
                    mEmailEt.setError("Invalid Email");
                    mEmailEt.setFocusable(true);
                } else {
                    //valid email pattern
                    loginUser(email, password);
                }
            }
        });

        notHaveAccountTv.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(LoginActivity.this, RegisterActivity.class));
                finish();
            }
        });

        //init progress dialog
        pd = new ProgressDialog(this);
        pd.setMessage("Logging In...");
    }

    private void loginUser(String email, String password) {
        //show progress dialog
        pd.show();
        mAuth.signInWithEmailAndPassword(email, password)
                    .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {
                            //dismiss progress dialog
                            pd.dismiss();

                            //add token to user
                            String current_user_id = mAuth.getCurrentUser().getUid();
                            String deviceToken = FirebaseInstanceId.getInstance().getToken();
                            mUserDatabase.child(current_user_id).child("device_token").setValue(deviceToken);

                            // Sign in success, update UI with the signed-in user's information
                            FirebaseUser user = mAuth.getCurrentUser();

                            //if user is signing in first time then get and show user info
                            if(task.getResult().getAdditionalUserInfo().isNewUser()){

                                //Get user email and uid from auth
                                String email = user.getEmail();
                                String uid = user.getUid();

                                //when user is registered store user info in firebase realtime database too
                                //using hashMap
                                HashMap<Object, String> hashMap = new HashMap<>();
                                hashMap.put("email", email);
                                hashMap.put("uid", uid);
                                hashMap.put("name", "");
                                hashMap.put("onlineStatus", "online");
                                hashMap.put("phone", ""); //will add later (e.g edit profile)
                                hashMap.put("image", ""); //will add later (e.g edit profile)
                                hashMap.put("cover", ""); //will add later (e.g edit profile)


                                //firebase database instance
                                FirebaseDatabase database = FirebaseDatabase.getInstance();

                                //path to store user data name "Users"
                                DatabaseReference reference = database.getReference("Users");

                                //put data within hashmap in database
                                reference.child(uid).setValue(hashMap);
                            }

                            //user is logged in, so start ProfileActivity
                            startActivity(new Intent(LoginActivity.this, DashboardActivity.class));
                            finish();
                        } else {
                            //dismiss progress dialog
                            pd.dismiss();
                            // If sign in fails, display a message to the user.
                            Toast.makeText(LoginActivity.this, "Authentication failed.",
                                    Toast.LENGTH_SHORT).show();

                        }

                    }
                }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception e) {
                //error, get and show error message
                Toast.makeText(LoginActivity.this, "" + e.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed(); // go previous activity
        return super.onSupportNavigateUp();
    }
}

File: Healthia/app/src/main/java/com/example/healthia/LoginActivity.kt
package com.example.healthia

import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.util.Patterns
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.example.healthia.LoginActivity
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import java.util.*

class LoginActivity : AppCompatActivity() {
    //views
    lateinit var mEmailEt: EditText
    lateinit var mPasswordEt: EditText
    lateinit var notHaveAccountTv: TextView
    lateinit var mLoginBtn: Button

    //Declare an instance of FirebaseAuth
    private var mAuth: FirebaseAuth? = null

    //progress dialog
    var pd: ProgressDialog? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        //actionbar and it's title
        val actionBar = supportActionBar
        actionBar!!.setTitle("Login")

        //enable back button
        actionBar.setDisplayHomeAsUpEnabled(true)
        actionBar.setDisplayShowHomeEnabled(true)

        //In the onCreate() method, initialize the FirebaseAuth instance.
        mAuth = FirebaseAuth.getInstance()

        //init
        mEmailEt = findViewById(R.id.emailEt)
        mPasswordEt = findViewById(R.id.passwordEt)
        notHaveAccountTv = findViewById(R.id.not_have_accountTv)
        mLoginBtn = findViewById(R.id.login_btn)

        //login button click
        mLoginBtn.setOnClickListener(View.OnClickListener {
            //input data
            val email = mEmailEt.getText().toString()
            val password = mPasswordEt.getText().toString().trim { it <= ' ' }
            if (!Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
                //invalid email pattern set error
                mEmailEt.setError("Invalid Email")
                mEmailEt.setFocusable(true)
            } else {
                //valid email pattern
                loginUser(email, password)
            }
        })
        notHaveAccountTv.setOnClickListener(View.OnClickListener { startActivity(Intent(this@LoginActivity, RegisterActivity::class.java)) })

        //init progress dialog
        pd = ProgressDialog(this)
        pd!!.setMessage("Logging In...")
    }

    private fun loginUser(email: String, password: String) {
        //show progress dialog
        pd!!.show()
        mAuth!!.signInWithEmailAndPassword(email, password)
                .addOnCompleteListener(this) { task ->
                    if (task.isSuccessful) {
                        //dismiss progress dialog
                        pd!!.dismiss()
                        // Sign in success, update UI with the signed-in user's information
                        val user = mAuth!!.currentUser

                        //if user is signing in first time then get and show user info
                        if (task.result!!.additionalUserInfo!!.isNewUser) {

                            //Get user email and uid from auth
                            val email = user!!.email
                            val uid = user.uid

                            //when user is registered store user info in firebase realtime database too
                            //using hashMap
                            val hashMap = HashMap<Any, String?>()
                            hashMap["email"] = email
                            hashMap["uid"] = uid
                            hashMap["name"] = ""
                            hashMap["onlineStatus"] = "online"
                            hashMap["phone"] = ""
                            hashMap["image"] = ""
                            hashMap["cover"] = ""


                            //firebase database instance
                            val database = FirebaseDatabase.getInstance()

                            //path to store user data name "Users"
                            val reference = database.getReference("Users")

                            //put data within hashmap in database
                            reference.child(uid).setValue(hashMap)
                        }

                        //user is logged in, so start ProfileActivity
                        startActivity(Intent(this@LoginActivity, MainActivity::class.java))
                        finish()
                    } else {
                        //dismiss progress dialog
                        pd!!.dismiss()
                        // If sign in fails, display a message to the user.
                        Toast.makeText(this@LoginActivity, "Authentication failed.",
                                Toast.LENGTH_SHORT).show()
                    }
                }.addOnFailureListener { e -> //error, get and show error message
                    Toast.makeText(this@LoginActivity, "" + e.message, Toast.LENGTH_SHORT).show()
                }
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed() // go previous activity
        return super.onSupportNavigateUp()
    }
}